packages/jupyter-ai/src/components/code-blocks/code-toolbar.tsx

```diff
@@ -0,0 +1,101 @@
+import React from 'react';
+import { Box } from '@mui/material';
+import { addAboveIcon, addBelowIcon } from '@jupyterlab/ui-components';
+
+import { CopyButton } from './copy-button';
+import { replaceCellIcon } from '../../icons';
+
+import {
+  ActiveCellManager,
+  useActiveCellContext
+} from '../../contexts/active-cell-context';
+import { TooltippedIconButton } from '../mui-extras/tooltipped-icon-button';
+
+export type CodeToolbarProps = {
+  /**
+   * The content of the Markdown code block this component is attached to.
+   */
+  content: string;
+};
+
+export function CodeToolbar(props: CodeToolbarProps): JSX.Element {
+  const [activeCellExists, activeCellManager] = useActiveCellContext();
+  const sharedToolbarButtonProps = {
+    content: props.content,
+    activeCellManager,
+    activeCellExists
+  };
+
+  return (
+    <Box
+      sx={{
+        display: 'flex',
+        justifyContent: 'flex-end',
+        alignItems: 'center',
+        padding: '6px 2px',
+        marginBottom: '1em',
+        border: '1px solid var(--jp-cell-editor-border-color)',
+        borderTop: 'none'
+      }}
+    >
+      <InsertAboveButton {...sharedToolbarButtonProps} />
+      <InsertBelowButton {...sharedToolbarButtonProps} />
+      <ReplaceButton {...sharedToolbarButtonProps} />
+      <CopyButton value={props.content} />
```

Comment:
Enhancement idea: any plan to let users customize the toolbar (including hiding some buttons or hiding the toolbar entirely)? A future enhancement might entail using the [toolbar registry](https://jupyterlab.readthedocs.io/en/stable/extension/extension_points.html#toolbar-registry) for this toolbar, as is used for other toolbars in Lab.

Reply:
Sure, that sounds like a good idea! This should be done in a future PR however, since adding a plugin API for this is a one-way decision and requires careful and deliberate design.